larpandoracontent/LArControlFlow/MasterAlgorithm.cc

```diff
@@ -418,6 +418,29 @@ StatusCode MasterAlgorithm::RunCosmicRayHitRemoval(const PfoList &ambiguousPfos)
     {
         const ClusterList clusterList(pPfoToDelete->GetClusterList());
         const VertexList vertexList(pPfoToDelete->GetVertexList());
+
+        // ATTN: If an ambiguous pfo has been stitched, reset the calo hit positions in preparation for subsequent algorithm chains
+        if (pPfoToDelete->GetPropertiesMap().count("X0"))
+        {
+            PfoList downstreamPfos;
+            LArPfoHelper::GetAllDownstreamPfos(pPfoToDelete, downstreamPfos);
```

Comment:
Are we doing this multiple times for some Pfos, given the call to `GetAllConnectedPfos` above, then `GetAllDownstreamPfos` here? Could use a PfoSet to keep track of which Pfos have been processed (but watch out, as these are unordered, so only insert when a Pfo has been processed, then look up to see whether to proceed with processing future Pfos).

Reply:
We don't actually do this multiple times because the X0 shift is only set for the parent cosmic ray and not the daughters.  This means that we need to reset the X0 position in the calo hits for the for the daughter pfos too when we encounter a pfo with an X0 because if we don't the daughter hits are still shifted.  

I believe the logic behind not setting the X0 for the daughters was that the shift is calculated from the parent track and so isn't a property of the daughters per say, but the daughter hits will be shifted because of it.  Perhaps it would be cleaner to set the x0 for the daughter pfos too as it is slightly confusing to shift the hits for the pfo, but not to set the x0.  What are your thoughts on this?  If you agree with that suggestions I'll add that another commit to this PR setting the X0 for the daughters.

Reply:
OK, thanks for clarifying. I think you have most experience now with where/how this is used, so please go with the convention you think makes most sense and will help ease maintenance in the future. If you want to bounce ideas around about this via e-mail or next week then I'd be very happy to discuss. Thanks.